src/sys/system_v/shm.rs

```diff
@@ -0,0 +1,476 @@
+        let size = std::mem::size_of::<T>();
+        // This is the main difference between this function and [`Shm::shmget`]
+        // Because we are always creating a new segment, we can be sure that the size match
+        let shmget_flag = ShmgetFlag::IPC_CREAT | ShmgetFlag::IPC_EXCL;
+        let flags = mode.bits() as i32 | shmget_flag.bits();
+        let id = Errno::result(unsafe { libc::shmget(key, size, flags) })?;
+        Ok(Self {
+            id,
+            _phantom: PhantomData,
+        })
+    }
+
+    /// Creates and returns a new, or returns an existing, System V shared memory
+    /// segment identifier.
+    ///
+    /// For more information, see [`shmget(2)`].
+    ///
+    /// # Safety
+    ///
+    /// If you are using this function to connect to an existing memory segment,
+    /// care must be taken that the generic type `T` matches what is actually
+    /// stored on the memory segment.\
+    /// For example, if a memory segment of size 4 bytes exist, and you connect
+    /// with a type of size 8 bytes, then undefined behaviour will be invoked.
+    ///
+    /// # Example
+    ///
+    /// ## Connecting to an existing shared memory segment
+    ///
+    /// ```no_run
+    /// # use nix::errno::Errno;
+    /// # use nix::sys::system_v::shm::*;
+    /// # use nix::sys::stat::Mode;
+    /// #
+    /// struct MyData(i64);
+    /// const MY_KEY: i32 = 1337;
+    ///
+    /// unsafe {
+    ///     let mem_segment = Shm::<MyData>::shmget(
+    ///         MY_KEY,
+    ///         ShmgetFlag::empty(),
+    ///         Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+    ///     )?;
+    /// }
+    /// # Ok::<(), Errno>(())
+    /// ```
+    ///
+    /// [`shmget(2)`]: https://man7.org/linux/man-pages/man2/shmget.2.html
+    pub unsafe fn shmget(
+        key: key_t,
+        shmget_flag: ShmgetFlag,
+        mode: Mode,
+    ) -> Result<Self> {
+        let size = std::mem::size_of::<T>();
+        let flags = mode.bits() as i32 | shmget_flag.bits();
+        let id = Errno::result(unsafe { libc::shmget(key, size, flags) })?;
+        Ok(Self {
+            id,
+            _phantom: PhantomData,
+        })
+    }
+
+    // -- Private --
+
+    /// Attaches the System V shared memory segment identified by `shmid` to the
+    /// address space of the calling process.
+    ///
+    /// This is called automatically on [`Shm::attach`].
+    ///
+    /// For more information, see [`shmat(2)`].
+    ///
+    /// [`shmat(2)`]: https://man7.org/linux/man-pages/man2/shmat.2.html
+    fn shmat(&self, shmat_flag: ShmatFlag) -> Result<*mut T> {
+        Errno::result(unsafe {
+            libc::shmat(self.id, ptr::null(), shmat_flag.bits())
+        })
+        .map(|ok| ok.cast::<T>())
+    }
+}
+
+#[derive(Debug)]
+/// Safe wrapper around a SystemV shared memory segment data
+///
+/// This is a smart pointer, and so implement the [`Deref`] and [`DerefMut`] traits.
```

Comment:
Nit: seems to be a mistake:

```suggestion
/// This is a smart pointer, and so implements the [`Deref`] and [`DerefMut`] traits.
```

Reply:
Fixed